### PR TITLE
Remove extra type aliases

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Follow-up of https://github.com/neondatabase/neon/pull/2231#discussion_r942313342

Removes both timeline-related type aliases, simplifying some type declarations:

Before: `local_timelines: HashMap<ZTimelineId, Arc<<RepositoryImpl as Repository>::Timeline>>,` 
After   : `local_timelines: HashMap<ZTimelineId, Arc<LayeredTimeline>>,`

### CODE REVIEW

**Comment by hlinnaka** [comment]
+1. I think I made most of these changes in https://github.com/neondatabase/neon/pull/2279 too

**Comment by hlinnaka** [comment]
> Before: local_timelines: HashMap<ZTimelineId, Arc<<RepositoryImpl as Repository>::Timeline>>,
> After : local_timelines: HashMap<ZTimelineId, Arc<LayeredTimeline>>,

+1 on the above, but I think the RepositoryImpl/TimelineImpl aliases still make the code nicer to read. It makes it more clear when code deals with an implementation of the Repository/Timeline trait, and isn't tied to the LayeredTimeline implementation. In practice, it compiles to the same of course.

Alternatively, we could rename `LayeredRepository` to `RepositoryImpl` and `LayeredTimeline` to `TimelineImpl`.

**Comment by SomeoneToIgnore** [comment]
> It makes it more clear when code deals with an implementation of the Repository/Timeline trait

I think I'm treating the traits differently, that's why reluctant to the idea: trait can be implemented for many things, so I never care if "N implements a trait", for me it's "a certain trait(s) is implemented for N" and that's why I'd like to avoid having "TraitImpl" in the struct name: we have two timeline-related traits already (albeit one extends another) and might have more.

So, I'd leave entities named as they are, but maybe @LizardWizzard can break a tie and find a compromise 🙂 
Either way, I'm fine with closing this and seeing an alternative version in the overhaul PR.

**Comment by LizardWizzard** [comment]
Hmm, I think I'm leaning towards @SomeoneToIgnore position. Repository/Timeline traits contain fair amount of implementation details of LayeredRepo and frequently we access Layered{Timeline, Repository} directly. Previously I said that it would be better to keep these traits as an API boundary, but now I think that we can achieve the same thing by properly using `pub` modifier (. Outside code wont be able to call internal methods, everything is simple. No need to workaround issues like [this](https://github.com/neondatabase/neon/pull/2279/files#diff-b14c4fdb873147f94591cf905d95eae74aa2d79d8c3e249c014db2bdfbe3c1caR359). Timeline and Repository traits look somewhat like a header file for the only implementation. So almost every time you add something new you need to add it to both places. 

To summarize I think it wont harm to remove these traits altogether and use `pub` to show which API's are expected to be used by other modules. This PR is a step in that direction.

Another thought is that Repository and Timeline carry too much weight, they're overloaded with different responsibilities, like gc, compaction stuff, datadir mapping etc. (Maybe this is only my feeling). So in the future I think we'll end up splitting them into several objects/modules with their own api boundaries. Though it would take some time to find the right balance. 

If we come up with another storage engine implementation it will also help to form better abstraction when it will be needed

Thoughts?

**Comment by hlinnaka** [comment]
> To summarize I think it wont harm to remove these traits altogether and use pub to show which API's are expected to be used by other modules. This PR is a step in that direction.

Ok, let's merge this.

**Review by hlinnaka — APPROVED** [review]
